Running command in Host fails when `/etc/iproute2/group` not present

[Here](https://gist.github.com/Infinoid/bf7e729025e38cf6979403e0fbcb6f9d) is a simple test program which creates a host and runs the command "ip addr" in it.

This succeeds when the file `/etc/iproute2/group` exists:

```console
% sudo touch /etc/iproute2/group     
% sudo go run .
ip addr said:
----
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 brd 127.255.255.255 scope global lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever

----
success
% 
```

And it fails when that file is missing.

```console
% sudo rm -v /etc/iproute2/group
removed '/etc/iproute2/group'
% sudo go run .                 
cmd.CombinedOutput() failed: exit status 2
failure
% 
```

If I ignore the error and print the combined output, I see a little more information:

```
panic: failed setup network bind mounts: failed to mount: no such file or directory

goroutine 1 [running]:
cunicu.li/gont/v2/pkg.init.0()
	/root/go/pkg/mod/cunicu.li/gont/v2@v2.5.0/pkg/exec.go:49 +0x185

```

And if I run it in strace, I see this sequence of syscalls:

```strace
newfstatat(AT_FDCWD, "/var/run/gont/testnet/files/etc/iproute2/.mount",  0xc0001e7078, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/var/run/gont/testnet/files/etc/iproute2/group", {st_mode=S_IFREG|0644, st_size=34, ...}, AT_SYMLINK_NOFOLLOW) = 0
mount("/var/run/gont/testnet/files/etc/hosts", "/etc/hosts", 0xc000130fab, MS_BIND, NULL) = 0
mount("/var/run/gont/testnet/files/etc/iproute2/group", "/etc/iproute2/group", 0xc000130fac, MS_BIND, NULL) = -1 ENOENT (No such file or directory)
```


Are you always running gont on hosts where that file is present?  Because it's not present here on this Debian box, and that seems to break command execution.  Even the `/etc/iproute2` folder didn't exist, I created it while tracking this down.

--- a/flake.nix
+++ b/flake.nix
@@ -8,23 +8,23 @@
     nixpkgs.url = "github:nixos/nixpkgs?ref=nixos-unstable";
   };
 
-  outputs = {
-    self,
-    flake-utils,
-    nixpkgs,
-  }:
-    flake-utils.lib.eachDefaultSystem
-    (
-      system: let
+  outputs =
+    {
+      self,
+      flake-utils,
+      nixpkgs,
+    }:
+    flake-utils.lib.eachDefaultSystem (
+      system:
+      let
         pkgs = nixpkgs.legacyPackages.${system};
-      in rec {
+      in
+      rec {
         packages.gont = pkgs.buildGoModule {
           name = "gont";
           src = ./.;
           vendorHash = "sha256-QOh1jBR7FL/fKFmJv7wGxuCghRLR3DV/0TzXd+bUFP0=";
-          buildInputs = with pkgs; [
-            libpcap
-          ];
+          buildInputs = with pkgs; [ libpcap ];
           doCheck = false;
         };
 
@@ -38,12 +38,10 @@
             packages.gont
           ];
 
-          inputsFrom = [
-            packages.gont
-          ];
+          inputsFrom = [ packages.gont ];
         };
 
-        formatter = nixpkgs.alejandra;
+        formatter = nixpkgs.nixfmt-rfc-style;
       }
     );
 }

--- a/internal/utils/utils.go
+++ b/internal/utils/utils.go
@@ -6,9 +6,14 @@ package utils
 import (
 	"math/rand"
 	"os"
+	"path/filepath"
 )
 
 func Touch(path string) error {
+	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
+		return err
+	}
+
 	f, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL, 0o444)
 	if err != nil {
 		return err

--- a/pkg/base_node.go
+++ b/pkg/base_node.go
@@ -73,7 +73,7 @@ func (n *Network) AddNode(name string, opts ...Option) (*BaseNode, error) {
 		}
 	}
 
-	// Create mount point dirs
+	// Create mount point directories
 	for _, ed := range node.EmptyDirs {
 		path := filepath.Join(basePath, "files", ed)
 

--- a/pkg/cmd_test.go
+++ b/pkg/cmd_test.go
@@ -13,6 +13,7 @@ import (
 
 	g "cunicu.li/gont/v2/pkg"
 	co "cunicu.li/gont/v2/pkg/options/cmd"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
@@ -151,4 +152,19 @@ func TestCmdContext(t *testing.T) {
 	require.True(t, ok)
 	require.True(t, ws.Signaled())
 	require.Equal(t, syscall.SIGKILL, ws.Signal())
+}
+
+func TestIProute2Files(t *testing.T) {
+	n, err := g.NewNetwork("")
+	require.NoError(t, err, "Failed to create network")
+	defer n.Close()
+
+	beep, err := n.AddHost("beep")
+	require.Nil(t, err)
+
+	cmd := beep.Command("ip", "addr")
+	out, err := cmd.CombinedOutput()
+	assert.Nil(t, err)
+
+	t.Logf("Output: %s", out)
 }

--- a/pkg/exec.go
+++ b/pkg/exec.go
@@ -4,6 +4,7 @@
 package gont
 
 import (
+	"errors"
 	"fmt"
 	"os"
 	"path/filepath"
@@ -11,6 +12,7 @@ import (
 	"syscall"
 
 	"cunicu.li/gont/v2/internal/execvpe"
+	"cunicu.li/gont/v2/internal/utils"
 	"golang.org/x/sys/unix"
 )
 
@@ -119,6 +121,32 @@ func setupBindMounts(basePath string) error {
 	for _, path := range files {
 		src := filepath.Join(filesRootPath, path)
 		tgt := filepath.Join("/", path)
+
+		srcInfo, err := os.Stat(src)
+		if err != nil {
+			return fmt.Errorf("failed to stat source: %s: %w", src, err)
+		}
+
+		// Create non-existing targets
+		if _, err := os.Stat(tgt); errors.Is(err, os.ErrNotExist) { //nolint:nestif
+			var tgtDir string
+			if srcInfo.IsDir() {
+				tgtDir = tgt
+			} else {
+				tgtDir = filepath.Dir(tgt)
+			}
+
+			// Create target if it does not exist yet
+			if srcInfo.IsDir() {
+				if err := os.MkdirAll(tgtDir, 0o755); err != nil {
+					return fmt.Errorf("failed to create directory: %s: %w", path, err)
+				}
+			} else {
+				if err := utils.Touch(tgt); err != nil {
+					return fmt.Errorf("failed to create empty file: %s: %w", tgt, err)
+				}
+			}
+		}
 
 		if err := syscall.Mount(src, tgt, "", syscall.MS_BIND, ""); err != nil {
 			return fmt.Errorf("failed to mount: %w", err)

--- a/pkg/nat_http_test.go
+++ b/pkg/nat_http_test.go
@@ -62,7 +62,7 @@ func TestGetMyIP(t *testing.T) {
 	require.NoError(t, err, "Failed to setup default route")
 
 	outp := &bytes.Buffer{}
-	_, err = client.Run("curl", "-sk", "--connect-timeout", 1000, "https://server",
+	_, err = client.Run("curl", "--silent", "--insecure", "--connect-timeout", 1000, "https://server",
 		co.Stdout(outp))
 	require.NoError(t, err, "Request failed")
 

--- a/pkg/network_files.go
+++ b/pkg/network_files.go
@@ -4,12 +4,16 @@
 package gont
 
 import (
+	"bufio"
 	"fmt"
 	"io"
 	"net"
 	"os"
 	"path/filepath"
+	"slices"
 	"strings"
+
+	"cunicu.li/gont/v2/internal/utils"
 )
 
 // IPv4loopback is the IPv4 loopback address (127.0.0.1)
@@ -90,10 +94,105 @@ func (n *Network) WriteHostsFile(f io.Writer) error {
 }
 
 func (n *Network) GenerateConfigFiles() error {
-	return n.GenerateIProute2Files()
+	if err := n.generateIProute2Files(); err != nil {
+		return err
+	}
+
+	// We need to patch /etc/nsswitch.conf here
+	// to avoid using systemd-resolved here as its
+	// broken from within network namesapces
+	if err := n.patchNSSConfFile(); err != nil {
+		return err
+	}
+
+	// We also need to hide the NSCD socket as glibc
+	// will otherwise use NSCD and hence again the resolve
+	// NSS module provided by systemd-resolved.
+	if err := n.hideNSCDSocket(); err != nil {
+		return err
+	}
+
+	return nil
 }
 
-func (n *Network) GenerateIProute2Files() error {
+func (n *Network) hideNSCDSocket() error {
+	fn := filepath.Join(n.VarPath, "files/var/run/nscd/socket")
+
+	// We hide the NSCD socket here by bind mounting
+	// an empty file over its location.
+	return utils.Touch(fn)
+}
+
+func readNSSwitchConfig(fn string) (map[string][]string, error) {
+	f, err := os.Open(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	m := map[string][]string{}
+	s := bufio.NewScanner(f)
+
+	for s.Scan() {
+		line := s.Text()
+
+		if strings.HasPrefix(line, "#") {
+			continue // Skip comments
+		}
+
+		cols := strings.Split(line, ":")
+
+		db := cols[0]
+		srcs := cols[1:]
+
+		m[db] = srcs
+	}
+
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func writeNSSwitchConfig(fn string, config map[string][]string) error {
+	f, err := os.OpenFile(fn, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if _, err = fmt.Fprintln(f, "# Gont's patched nsswitch.conf"); err != nil {
+		return err
+	}
+
+	for db, srcs := range config {
+		if _, err := fmt.Fprintf(f, "%s: %s\n", db, strings.Join(srcs, " ")); err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+func (n *Network) patchNSSConfFile() error {
+	cfg, err := readNSSwitchConfig("/etc/nsswitch.conf")
+	if err != nil {
+		return fmt.Errorf("failed to read nsswitch.conf: %w", err)
+	}
+
+	for db := range cfg {
+		if db == "hosts" {
+			cfg[db] = slices.DeleteFunc(cfg[db], func(src string) bool {
+				return !strings.HasPrefix(src, "resolve") && !strings.HasPrefix(src, "mymachines") && !strings.HasPrefix(src, "myhostname")
+			})
+		}
+	}
+
+	fn := filepath.Join(n.VarPath, "files/etc/nsswitch.conf")
+	return writeNSSwitchConfig(fn, cfg)
+}
+
+func (n *Network) generateIProute2Files() error {
 	fn := filepath.Join(n.VarPath, "files/etc/iproute2/group")
 	if err := os.MkdirAll(filepath.Dir(fn), 0o755); err != nil {
 		return err

--- a/pkg/options/cmd/exec.go
+++ b/pkg/options/cmd/exec.go
@@ -75,9 +75,15 @@ func (e Env) ApplyExecCmd(c *exec.Cmd) {
 	c.Env = append(c.Env, string(e))
 }
 
-// EnvVar appends a key-value paired environment variable
+// EnvVar appends a key-value paired environment variable.
 func EnvVar(k, v string) Env {
 	return Env(fmt.Sprintf("%s=%s", k, v))
+}
+
+// PassEnv forward environment variables from the main
+// Gont process to the invoked sub-processes.
+func PassEnv(k string) Env {
+	return EnvVar(k, os.Getenv(k))
 }
 
 // Envs appends additional environment variables.

